modules/ocf/files/ssl/ocf-lets-encrypt

```diff
@@ -0,0 +1,254 @@
+    0       nothing changed
+    255     cert renewed/acquired
+    else    error
+"""
+import argparse
+import functools
+import os
+import re
+import socket
+import subprocess
+import sys
+import tempfile
+from contextlib import contextmanager
+from datetime import datetime
+from datetime import timedelta
+from datetime import timezone
+from pathlib import Path
+
+import dateutil.parser
+from OpenSSL import crypto
+
+
+RENEW_WHEN_TIME_REMAINING = timedelta(days=45)
+
+LE_ACCOUNT_KEY = Path('/etc/ssl/lets-encrypt/le-account.key')
+ACME_BASE_DIR = Path('/var/lib/lets-encrypt/')
+
+
+def debug(*args, **kwargs):
+    pass
+
+
+def expiration_date(cert_path):
+    with cert_path.open('rb') as f:
+        cert = crypto.load_certificate(crypto.FILETYPE_PEM, f.read())
+    return dateutil.parser.parse(cert.get_notAfter())
+
+
+def cert_status(cert_path):
+    if cert_path.exists():
+        expires = expiration_date(cert_path)
+        time_remaining = expires - datetime.now(timezone.utc)
+        if time_remaining > RENEW_WHEN_TIME_REMAINING:
+            return False, time_remaining
+        else:
+            return True, time_remaining
+    else:
+        return True, None
+
+
+@contextmanager
+def make_csr(domains, key_path):
+    req = crypto.X509Req()
+    req.get_subject().CN = domains[0]
+
+    with open(key_path, 'rt') as f:
+        key = crypto.load_privatekey(crypto.FILETYPE_PEM, f.read())
+        req.set_pubkey(key)
+
+    req.add_extensions((crypto.X509Extension(
+        b'subjectAltName',
+        False,
+        bytes(', '.join('DNS:' + domain for domain in domains), encoding='ascii'),
+    ),))
+
+    req.sign(key, 'sha256')
+
+    with tempfile.NamedTemporaryFile() as csr:
+        csr.write(crypto.dump_certificate_request(crypto.FILETYPE_PEM, req))
+        csr.flush()
+        yield csr.name
+
+
+def acme_tiny(domains, key_path):
+    with make_csr(domains, key_path) as csr_path:
+        # TODO: use subprocess.run when we get python3.5
+        proc = subprocess.Popen(
+            (
+                'acme-tiny',
+                '--account-key', str(LE_ACCOUNT_KEY),
+                '--csr', csr_path,
+                '--acme-dir', str(ACME_BASE_DIR / '.well-known' / 'acme-challenge'),
+                # '--ca', 'https://acme-staging.api.letsencrypt.org',
+            ),
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        stdout, stderr = proc.communicate()
+    return stdout, stderr, proc.returncode
+
+
+def write_cert(cert, cert_path):
+    """Write certificate file."""
+    with tempfile.NamedTemporaryFile(
+        dir=str(cert_path.parent),
+        prefix='.tmp',
+        delete=False,
+    ) as temp_cert:
+        try:
+            temp_cert.write(cert)
+            # atomic overwrite in case of errors
+            os.chmod(temp_cert.name, 0o644)
+            os.rename(temp_cert.name, str(cert_path))
+        finally:
+            try:
+                os.remove(temp_cert.name)
```

Comment:
Eh, fwiw if it were me, I would only remove the file if the rename did not succeed. This would avert the small chance that some other process created a file with the same name as `temp_cert.name` between a successful rename and a remove. That having been said, it's a mostly theoretical race...

Reply:
Meh, somehow I think it's more likely for `os.rename` to mysteriously silently fail and not trigger the remove than for this race to happen.